File: browser-sources/boss.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
browser-sources/boss.py — BOSS直聘 (zhipin.com) login-state source, nodriver sidecar.

The BROWSER half of the boss source. The Node wrapper (`browser-sources/boss.mjs`)
spawns this script and applies title/location filters + dedup + pipeline writes
in JS (reusing scan.mjs's pure functions), so the two halves stay byte-compatible
with the zhaopin source. This script only: login handoff, search, extract the
Vue `jobList` state (plaintext salary — the DOM renders salary with a custom
obfuscation font, but the underlying Vue data is plain), and emit raw jobs as
NDJSON.

Why nodriver and not Playwright: BOSS risk control rejects Playwright/CDP-driven
browsers (even headed real Chrome, with a valid session) — about:blank + reload
loop / window.close(). nodriver drives real Chrome without exposing automation
fingerprints (navigator.webdriver=False). See docs/dev-team/carreer-ops/recon-report.md.

Hard discipline (QA acceptance, mandatory):
  - single page at a time, no auto-retry;
  - page cap (--max-pages, default 5);
  - abort on risk signature (login wall / reload loop), never silent 0;
  - graceful close (Browser.close) so the session persists in the profile.

Usage:
  python boss.py --login                                   # handoff: sign in once
  python boss.py --jsonl --keywords 数据分析,BI --max-pages 5
      # NDJSON: {keyword, rawJobs:[...], errors:[...], done, total} per keyword,
      # then {done:true, source:"boss", total}
"""
import argparse, asyncio, json, os, pathlib, subprocess, sys, time, urllib.request
import nodriver as uc
from nodriver.cdp import browser as cdp_browser

# ── single authoritative path resolution ─────────────────────────────────
# This script lives at <repo-root>/browser-sources/boss.py (a FIXED, documented
# location), so repo root = parents[1]. Env override for portability/tests.
ROOT = pathlib.Path(os.environ.get("CAREER_OPS_ROOT", pathlib.Path(__file__).resolve().parent.parent))
PROFILE = pathlib.Path(os.environ.get("BOSS_PROFILE", ROOT / "config" / "browser-state" / "boss-profile"))
CHROME_CANDIDATES = [
    "C:/Program Files/Google/Chrome/Application/chrome.exe",
    "C:/Program Files (x86)/Google/Chrome/Application/chrome.exe",
]
DEFAULT_CITY = "101020100"  # 上海
PAGE_SIZE = 15
BASE_URL = "https://www.zhipin.com/web/geek/job"

def val(x):
    return getattr(x, "value", x)

def find_chrome():
    for p in CHROME_CANDIDATES:
        if pathlib.Path(p).exists():
            return p
    return None

# ── single browser lifecycle (the shared :9222 debug Chrome) ──────────────
# Every boss flow — parse-link, scan, login handoff — attaches to ONE Chrome
# instance on a fixed debug port, reusing the boss-profile dir. Nodriver's
# `uc.start(user_data_dir=…)` would otherwise spawn a SECOND Chrome on the
# same profile; two Chromes cannot share a profile dir (singleton lock), so
# parse-link's :9222 instance and a scan's fresh instance collided. Launching
# goes through spawn_debug_chrome() here (idempotent — attaches when :9222 is
# already up), mirroring web/src/lib/boss-chrome.ts's getChromeCdpUrl().
DEBUG_PORT = 9222
LANDING_URL = f"{BASE_URL}?query=%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90&city={DEFAULT_CITY}"


def chrome_cdp_ready():
    try:
        with urllib.request.urlopen(f"http://127.0.0.1:{DEBUG_PORT}/json/version", timeout=1.0) as r:
            return r.status == 200
    except Exception:
        return False


def spawn_debug_chrome():
    chrome = find_chrome()
    if not chrome:
        return False
    args = [
        chrome,
        f"--remote-debugging-port={DEBUG_PORT}",
        f"--user-data-dir={PROFILE}",
        "--start-maximized",
        LANDING_URL,
    ]
    try:
        subprocess.Popen(
            args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            stdin=subprocess.DEVNULL,
            creationflags=getattr(subprocess, "DETACHED_PROCESS", 0)
            | getattr(subprocess, "CREATE_NEW_PROCESS_GROUP", 0),
        )
    except Exception:
        return False
    for _ in range(40):  # up to ~20s for the DevTools port to come up
        if chrome_cdp_ready():
            return True
        time.sleep(0.5)
    return False


async def ensure_browser():
    """Attach to the shared :9222 debug Chrome, launching it first if needed.
    Returns the nodriver Browser, or None when Chrome is unavailable."""
    if not chrome_cdp_ready():
        if not spawn_debug_chrome():
            return None
    try:
        return await uc.start(host="127.0.0.1", port=DEBUG_PORT)
    except Exception:
        return None

def write_marker():
    """Touch config/browser-state/boss.json so the web Explorer's fail-fast
    login-state check (loginStatePath) passes, and record the profile path."""
    marker = ROOT / "config" / "browser-state" / "boss.json"
    try:
        marker.parent.mkdir(parents=True, exist_ok=True)
        marker.write_text(json.dumps({"profile": str(PROFILE), "updated": time.time()}), encoding="utf-8")
    except Exception:
        pass

async def nav_count(tab):
    v = val(await tab.evaluate("performance.getEntriesByType('navigation').length", return_by_value=True))
    try:
        return int(v)
    except Exception:
        return 0

def risk_reason(url):
    if not url:
        return "no URL"
    if "passport" in url or "verify" in url or "security" in url or "captcha" in url:
        return f"risk wall: {url[:80]}"
    if url.startswith("about:blank"):
        return "about:blank (reload-loop signature)"
    return None

def logged_in(url):
    return url and "passport" not in url and "verify" not in url and "/web/user/" not in url and "/web/geek" in url

async def read_state(tab):
    """Read the search page's Vue jobList (plaintext) + pagination flags."""
    js = r"""
    (function(){
      var cards = document.querySelectorAll('[class*=job-card]');
      var list = [];
      if (cards.length) {
        var v = cards[0].__vue__;
        var node = v, depth = 0;
        while (node && depth < 8) {
          var d = node.$data || node._data || {};
          if (Array.isArray(d.jobList)) { list = d.jobList; return JSON.stringify({
            jobList: list, hasMore: !!d.hasMore, page: (d.pageVo||{}).page, pageSize: (d.pageVo||{}).pageSize
          }); }
          node = node.$parent; depth++;
        }
      }
      return JSON.stringify({jobList: [], hasMore: false, page: null, pageSize: null, noState: true});
    })()
    """
    raw = val(await tab.evaluate(js, return_by_value=True)) or "{}"
    try:
        return json.loads(raw)
    except Exception:
        return {"jobList": [], "hasMore": False, "page": None, "pageSize": None, "parseError": True}

def normalize(raw):
    """raw Vue job object -> raw job dict (salary stays a string; JS parses it)."""
    if not raw or not isinstance(raw, dict):
        return None
    title = str(raw.get("jobName") or "").strip()
    eid = str(raw.get("encryptJobId") or "").strip()
    if not title or not eid:
        return None
    loc = " ".join(x for x in (raw.get("cityName"), raw.get("areaDistrict"), raw.get("businessDistrict"))
                   if isinstance(x, str) and x.strip())
    return {
        "title": title,
        "url": f"https://www.zhipin.com/job_detail/{eid}.html",
        "company": str(raw.get("brandName") or "").strip(),
        "location": loc,
        "salaryDesc": str(raw.get("salaryDesc") or "").strip(),
        "experience": str(raw.get("jobExperience") or "").strip(),
        "degree": str(raw.get("jobDegree") or "").strip(),
    }

async def scan_keyword(tab, kw, max_pages):
    """Load a keyword's results and collect raw jobs across up to max_pages
    scroll rounds. Returns (jobs, error)."""
    url = f"{BASE_URL}?query={kw}&city={DEFAULT_CITY}"
    jobs = []
    seen = set()
    try:
        await tab.get(url)
        await tab.sleep(6)
        # reload-loop detection (hard discipline): >2 navigations in the next
        # 7s = risk signature → abort, never silently 0.
        n0 = await nav_count(tab)
        await tab.sleep(7)
        n1 = await nav_count(tab)
        cur = val(await tab.evaluate("location.href", return_by_value=True)) or ""
        rrisk = risk_reason(cur)
        if rrisk:
            return [], rrisk
        if n1 - n0 > 2:
            return [], f"reload-loop signature ({n1 - n0} navigations/7s) — aborting"
        for page in range(max_pages):
            state = await read_state(tab)
            if state.get("noState") or state.get("parseError"):
                if not jobs:
                    return [], "Vue state not found (page structure changed?) — inspect live page"
                break
            for raw in state.get("jobList") or []:
                j = normalize(raw)
                if j and j["url"] not in seen:
                    seen.add(j["url"])
                    jobs.append(j)
            if not state.get("hasMore"):
                break
            # load more: scroll to bottom (BOSS search page scrolls to load).
            await tab.evaluate("window.scrollTo(0, document.documentElement.scrollHeight)")
            await tab.sleep(3)
        return jobs, None
    except Exception as e:
        return jobs, f"{type(e).__name__}: {e}"

async def login_handoff():
    browser = await ensure_browser()
    if browser is None:
        print("BROWSER_FAILED: 无法启动 BOSS 调试浏览器（未找到 Chrome 或启动失败）", flush=True)
        return
    try:
        tab = await browser.get(LANDING_URL)
        print("[boss] 请在弹出的 Chrome 窗口完成 BOSS直聘 登录；登录成功跳回职位页后自动保存。", flush=True)
        deadline = time.time() + 12 * 60
        while time.time() < deadline:
            await tab.sleep(4)
            url = val(await tab.evaluate("location.href", return_by_value=True)) or ""
            if logged_in(url):
                write_marker()
                print("LOGIN_SAVED " + str(PROFILE), flush=True)
                break
        else:
            print("LOGIN_TIMEOUT", flush=True)
    finally:
        try:
            await browser.stop()
        except Exception:
            pass
        try:
            await browser.aclose()
        except Exception:
            pass

async def run_scan(keywords, max_pages):
    browser = await ensure_browser()
    if browser is None:
        print(json.dumps({"done": True, "source": "boss", "total": 0, "errors": ["browser unavailable — Chrome not found or failed to start"]}, ensure_ascii=True), flush=True)
        return
    total = 0
    try:
        tab = await browser.get("about:blank", new_tab=True)
        for i, kw in enumerate(keywords):
            url = val(await tab.evaluate("location.href", return_by_value=True)) or ""
            if not logged_in(url):
                # fresh profile: prompt login via the search URL so the user can sign in
                await tab.get(f"{BASE_URL}?query={kw}&city={DEFAULT_CITY}")
                await tab.sleep(3)
                url = val(await tab.evaluate("location.href", return_by_value=True)) or ""
                if not logged_in(url):
                    print(json.dumps({"keyword": kw, "rawJobs": [], "errors": ["not logged in — run `node scan-browser-source.mjs boss --login` first"], "done": i + 1, "total": len(keywords)}), flush=True)
                    continue
            jobs, err = await scan_keyword(tab, kw, max_pages)
            total += len(jobs)
            print(json.dumps({"keyword": kw, "rawJobs": jobs, "errors": [err] if err else [], "done": i + 1, "total": len(keywords)}), flush=True)
        write_marker()
        print(json.dumps({"done": True, "source": "boss", "total": total}), flush=True)
    finally:
        try:
            await browser.stop()
        except Exception:
            pass
        try:
            await browser.aclose()
        except Exception:
            pass

async def parse_link(url):
    """Attach to the shared :9222 debug Chrome (launching it if needed) and
    evaluate the BOSS job_detail page in a throwaway tab. Returns a structured
    browserNotRunning only when the browser can't be started at all.

    Contract: this function ALWAYS prints exactly one JSON line —
    {documentTitle,metaDesc,locationEl} | {loginRequired:true} |
    {browserNotRunning:true} | {error:msg}. The web route keys off these."""
    browser = await ensure_browser()
    if browser is None:
        print(json.dumps({"browserNotRunning": True, "error": "Chrome 未在 :9222 运行且自动启动失败"}, ensure_ascii=True), flush=True)
        return
    tab = None
    try:
        # new_tab=True is mandatory in attach mode: bare browser.get() REUSES
        # the browser's first tab, and tab.close() below would then close the
        # user's own tab — closing their last tab kills the whole shared
        # Chrome (observed: browser gone after one parse).
        tab = await browser.get(url, new_tab=True)
        await tab.sleep(10)
        js = r"""
        (function(){
          var desc = (document.querySelector('meta[name=description]') || {}).content || '';
          var el = document.querySelector('.company-location');
          return JSON.stringify({
            documentTitle: document.title || '',
            metaDesc: desc,
            locationEl: el ? el.textContent.trim() : ''
          });
        })()
        """
        raw = val(await tab.evaluate(js, return_by_value=True)) or "{}"
        try:
            data = json.loads(raw)
        except Exception:
            data = {"title": "", "company": "", "location": "", "error": "parse failed"}
        login_js = r"""
        (function(){
          // Match the PATH, not the full href: job_detail links legitimately
          // carry ?securityId=… in the query, and matching 'security' against
          // the whole URL false-positives every shared link as a login wall
          // even when the detail page rendered fine (observed 2026-09-08).
          var p = (location.pathname || '').toLowerCase();
          var t = (document.title || '').toLowerCase();
          var wall = /login|security|passport|register|verify|captcha/.test(p) || /登录|注册/.test(t);
          var body = (document.body && document.body.innerText) ? document.body.innerText.slice(0,120) : '';
          if (!wall) wall = /登录|注册/.test(body) && !/(数据分析|产品|工程师|助理|专员|经理|主管|运营|市场|销售|财务|人力|行政|设计|开发|java|python|前端|后端|算法|数据)/.test(body);
          return JSON.stringify({ loginRequired: !!wall, url: location.href, title: document.title || '' });
        })()
        """
        login_raw = val(await tab.evaluate(login_js, return_by_value=True)) or "{}"
        try:
            login_info = json.loads(login_raw)
        except Exception:
            login_info = {}
        if login_info.get("loginRequired"):
            print(json.dumps({"loginRequired": True, "url": login_info.get("url", ""), "title": login_info.get("title", "")}, ensure_ascii=True), flush=True)
        else:
            print(json.dumps(data, ensure_ascii=True), flush=True)
    except Exception as e:
        # browser.get/evaluate failures (dead target, navigation abort, …) must
        # still surface as structured JSON — a bare traceback leaves stdout
        # empty and the web route can only show a misleading generic error.
        print(json.dumps({"error": f"{type(e).__name__}: {e}"}, ensure_ascii=True), flush=True)
    finally:
        if tab is not None:
            try:
                await tab.close()
            except Exception:
                pass

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--login", action="store_true")
    ap.add_argument("--jsonl", action="store_true")
    ap.add_argument("--keywords", default="")
    ap.add_argument("--max-pages", type=int, default=5)
    ap.add_argument("--parse-link", default="")
    args = ap.parse_args()

    if args.login:
        asyncio.run(login_handoff())
        return
    if args.parse_link:
        # Last-resort guard: stdout must always carry exactly one JSON line,
        # even if the event loop itself blows up (RuntimeError at teardown…).
        try:
            asyncio.run(parse_link(args.parse_link))
        except Exception as e:
            print(json.dumps({"error": f"parse-link crashed: {type(e).__name__}: {e}"}, ensure_ascii=True), flush=True)
        return

    kws = [k.strip() for k in args.keywords.split(",") if k.strip()] if args.keywords else ["数据分析", "BI", "用户运营"]
    asyncio.run(run_scan(kws, args.max_pages))

if __name__ == "__main__":
    main()

File: browser-sources/boss.mjs
// @ts-check
/**
 * browser-sources/boss.mjs — BOSS直聘 (zhipin.com) login-state source.
 *
 * The NODE half of the boss source. The browser work (login handoff + search +
 * extraction) is done by the nodriver Python sidecar `browser-sources/boss.py`
 * — BOSS risk control rejects Playwright/CDP-driven browsers, so the base
 * `BrowserSource` engine (Playwright) cannot be used here. This module keeps
 * the SAME contract as every other source (id / label / run / login), spawns
 * the sidecar, and applies title/location filter + dedup + pipeline writes in
 * JS by reusing scan.mjs's pure functions — so output is byte-identical to the
 * zhaopin source.
 *
 * See docs/dev-team/carreer-ops/recon-report.md for the full recon trail
 * (nodriver rationale, Vue `jobList` plaintext schema, R4 code table, hard
 * discipline, account-residual-risk history).
 */
import { spawn } from 'child_process';
import { existsSync, readFileSync } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import * as yaml from 'js-yaml';
import {
  appendToPipeline,
  appendToScanHistory,
  loadSeenUrls,
  buildLocationFilter,
} from '../scan.mjs';
import { buildTitleFilter } from '../lib/browser-source.mjs';
import { localToday } from '../lib/local-today.mjs';
import { resolvePythonExecutable } from '../web/src/lib/python-env.mjs';

const ROOT = path.resolve(fileURLToPath(new URL('..', import.meta.url)));
const PORTALS_PATH = path.join(ROOT, 'portals.yml');
const BOSS_PY = path.join(ROOT, 'browser-sources', 'boss.py');
const DEFAULT_CITY = '101020100'; // 上海

// ── pure helpers (exported for tests, mirror zhaopin.mjs) ────────────────

/**
 * Build the zhipin geek search URL. BOSS paginates client-side (pageVo.page +
 * hasMore); the `page` URL param is accepted but the SPA drives paging itself.
 * @param {string} keyword
 * @param {number} [page]
 * @param {string} [cityCode]
 * @returns {string}
 */
export function buildSearchUrl(keyword, page = 1, cityCode = DEFAULT_CITY) {
  const q = new URLSearchParams({ query: keyword, city: cityCode });
  if (page > 1) q.set('page', String(page));
  return `https://www.zhipin.com/web/geek/job?${q.toString()}`;
}

/**
 * Parse a BOSS salary display ("18-35K", "18-35K·13薪", "30K以上") into the
 * pipeline compensation contract. The ·N薪 bonus suffix does not change the
 * monthly range. Unknown shapes → null (drop the field, stay broad).
 * @param {unknown} value
 * @returns {{min:number,max:number,currency:string}|null}
 */
export function parseSalary(value) {
  const s = String(value ?? '').trim();
  if (!s || s === '面议' || s === '薪资面议') return null;
  const norm = s
    .replace(/·\s*\d+\s*薪/g, '')
    .replace(/k/g, 'K')
    .replace(/\s+/g, '');
  const range = norm.match(/^(\d+(?:\.\d+)?)\s*-\s*(\d+(?:\.\d+)?)K$/);
  if (range) {
    const min = Math.round(Number(range[1]) * 1000);
    const max = Math.round(Number(range[2]) * 1000);
    if (min > 0 && max >= min) return { min, max, currency: 'CNY' };
  }
  const above = norm.match(/^(\d+(?:\.\d+)?)K以上$/);
  if (above) {
    const min = Math.round(Number(above[1]) * 1000);
    // "以上" is open-ended; collapsed to min=max (no invented upper bound) —
    // consistent with zhaopin's parseSalary single-number convention.
    if (min > 0) return { min, max: min, currency: 'CNY' };
  }
  const single = norm.match(/^(\d+(?:\.\d+)?)K$/);
  if (single) {
    const n = Math.round(Number(single[1]) * 1000);
    if (n > 0) return { min: n, max: n, currency: 'CNY' };
  }
  return null;
}

/**
 * Normalize a raw Vue job object (from boss.py) into the pipeline Job shape.
 * Pure — exported for tests.
 * @param {object} raw
 * @returns {{title:string,url:string,company:string,location:string,salary?:{min:number,max:number,currency:string}}|null}
 */
export function normalizeJob(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const title = String(raw.title ?? '').trim();
  const url = String(raw.url ?? '').trim();
  if (!title || !/^https?:\/\//i.test(url)) return null;
  const out = {
    title,
    url: url.replace(/^http:\/\//i, 'https://'),
    company: String(raw.company ?? '').trim(),
    location: String(raw.location ?? '').trim(),
  };
  const salary = parseSalary(raw.salaryDesc);
  if (salary) out.salary = salary;
  return out;
}

/** Load portals.yml (or the ephemeral CAREER_OPS_PORTALS) tolerantly. */
function loadPortals() {
  const portalPath = process.env.CAREER_OPS_PORTALS || PORTALS_PATH;
  if (!existsSync(portalPath)) return {};
  try {
    const doc = yaml.load(readFileSync(portalPath, 'utf8'));
    return doc && typeof doc === 'object' ? doc : {};
  } catch {
    return {};
  }
}

// ── sidecar bridge ───────────────────────────────────────────────────────

/**
 * Spawn the Python sidecar and stream its NDJSON stdout lines to `onLine`.
 * @param {string[]} args
 * @param {(line: string) => void} onLine
 * @returns {Promise<{code: number|null, stderr: string}>}
 */
function spawnSidecar(args, onLine) {
  return new Promise((resolve) => {
    // Resolve the interpreter explicitly (D-006): an Explorer/cmd-started
    // parent can have a PATH whose `python` lacks nodriver (e.g. the
    // agent-reach venv), killing boss.py on import with no stdout.
    const pythonExe = resolvePythonExecutable({ requireModule: 'nodriver' });
    if (!pythonExe) {
      resolve({ code: 1, stderr: 'no python with nodriver found — set CAREER_OPS_PYTHON to the boss.py interpreter' });
      return;
    }
    const child = spawn(pythonExe, [BOSS_PY, ...args], {
      cwd: ROOT,
      env: { ...process.env, PYTHONIOENCODING: 'utf-8' },
      stdio: ['ignore', 'pipe', 'pipe'],
    });
    let buf = '';
    let stderr = '';
    child.stdout.on('data', (d) => {
      buf += d.toString('utf-8');
      let i;
      while ((i = buf.indexOf('\n')) !== -1) {
        const line = buf.slice(0, i).trim();
        buf = buf.slice(i + 1);
        if (line) onLine(line);
      }
    });
    child.stderr.on('data', (d) => { stderr += d.toString('utf-8'); });
    child.on('close', (code) => resolve({ code, stderr }));
  });
}

/** @type {import('../lib/browser-source.mjs').BrowserSource & object} */
const boss = {
  id: 'boss',
  label: 'BOSS直聘 (zhipin.com)',
  loginUrl: 'https://www.zhipin.com/web/geek/job',
  defaultKeywords: ['数据分析', 'BI', '用户运营'],
  configSection: 'boss_searches',

  /** Open a nodriver Chrome window for the user to sign in (handoff). */
  async login() {
    const { code, stderr } = await spawnSidecar(['--login'], () => {});
    if (code !== 0) console.error(`boss --login exited ${code}: ${stderr}`);
  },

  /**
   * Run the scan: spawn the sidecar per keyword stream, filter + dedup in JS,
   * emit the same NDJSON the web Explorer expects, and write pipeline/history.
   */
  async run({ dryRun = false, keyword = null, keywords = null, json = false, onKeyword = null, maxPages = 5 } = {}) {
    const portals = loadPortals();
    const section = portals[this.configSection];
    const kwList = (Array.isArray(keywords) && keywords.length)
      ? keywords
      : (keyword ? [keyword]
        : (Array.isArray(section) && section.length
          ? section.map((e) => (typeof e === 'string' ? e : e?.kw ?? e?.was)).filter((k) => typeof k === 'string' && k.trim())
          : this.defaultKeywords));

    const matchesTitle = buildTitleFilter(portals.title_filter);
    const matchesLocation = buildLocationFilter(portals.location_filter);
    const { seen } = loadSeenUrls();
    const date = localToday();

    const newOffers = [];
    const titleSkipped = [];
    const locationSkipped = [];
    const dupeSkipped = [];
    const errors = [];

    const handleLine = (line) => {
      let frame;
      try { frame = JSON.parse(line); } catch { return; }
      if (frame.done === true && frame.source === 'boss') return; // terminal frame
      const kw = frame.keyword ?? '';
      const rawJobs = Array.isArray(frame.rawJobs) ? frame.rawJobs : [];
      const kwErrors = Array.isArray(frame.errors) ? frame.errors : [];
      const kwOffers = [];
      for (const raw of rawJobs) {
        const job = normalizeJob(raw);
        if (!job) continue;
        job.source = this.id;
        if (!matchesTitle(job.title)) { seen.add(job.url); titleSkipped.push(job); continue; }
        if (!matchesLocation(job.location, job.url, job.title)) { seen.add(job.url); locationSkipped.push(job); continue; }
        if (seen.has(job.url)) { dupeSkipped.push(job); continue; }
        seen.add(job.url);
        newOffers.push(job);
        kwOffers.push(job);
      }
      for (const e of kwErrors) errors.push({ keyword: kw, error: e });
      if (onKeyword) onKeyword(kw, { offers: kwOffers, errors: kwErrors.map((e) => ({ keyword: kw, error: e })), done: frame.done, total: frame.total });
    };

    // O2 fail-fast (matches the web loginStatePath check): without the marker
    // written by boss.py on a successful login/scan, spawn nothing — a headed
    // window would just report "not logged in" after a pointless browser start.
    const markerPath = path.join(ROOT, 'config', 'browser-state', 'boss.json');
    if (!existsSync(markerPath)) {
      const err = `未找到登录态，请先运行 node scan-browser-source.mjs boss --login 登录后重试`;
      if (onKeyword) onKeyword(kwList[0] ?? '', { offers: [], errors: [{ keyword: kwList[0] ?? '', error: err }], done: 0, total: kwList.length });
      errors.push({ keyword: '', error: err });
      return { newOffers: [], titleSkipped: [], locationSkipped: [], dupeSkipped: [], errors, keywords: kwList, date };
    }

    const { code, stderr } = await spawnSidecar(
      ['--jsonl', '--keywords', kwList.join(','), '--max-pages', String(maxPages)],
      handleLine,
    );
    if (code !== 0 && stderr) console.error(`boss sidecar stderr: ${stderr}`);

    if (!dryRun) {
      if (newOffers.length) await appendToPipeline(newOffers);
      if (newOffers.length) await appendToScanHistory(newOffers, date, 'added');
      if (titleSkipped.length) await appendToScanHistory(titleSkipped, date, 'skipped_title');
      if (locationSkipped.length) await appendToScanHistory(locationSkipped, date, 'skipped_location');
      if (dupeSkipped.length) await appendToScanHistory(dupeSkipped, date, 'skipped_dup');
    }

    return { newOffers, titleSkipped, locationSkipped, dupeSkipped, errors, keywords: kwList, date };
  },
};

export default boss;
